File: 24-swap-nodes-in-pairs/swap-nodes-in-pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode* dummy = new ListNode(0, head);
        ListNode* p = dummy;
        while(p->next && p->next->next) {
            ListNode* l = p->next;
            ListNode* r = p->next->next;
            l->next = r->next;
            r->next = l;
            p->next = r;
            p = p->next->next;
        }
        return dummy->next;
    }
};

File: 24-swap-nodes-in-pairs/swap-nodes-in-pairs.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
    dummy := &ListNode{
        Next: head,
    }
    p := dummy
    for p.Next != nil && p.Next.Next != nil {
        l := p.Next
        r := p.Next.Next
        l.Next = r.Next
        r.Next = l
        p.Next = r
        p = p.Next.Next
    }

    return dummy.Next
}